book/src/implemented-proposals/repair-service.md

```diff
@@ -2,48 +2,96 @@
+will halt replay progress on this node.
+
+## Repair-related primitives
+Epoch Slots: 
+   Each validator advertises separately on gossip thhe various parts of an 
+   `Epoch Slots`:
+   * The `stash`: An epoch-long compressed set of all completed slots.
+   * The `cache`: The Run-length Encoding (RLE) of the latest `N` completed
+     slots starting from some some slot `M`, where `N` is the number of slots
+     that will fit in an MTU-sized packet.
+
+   `Epoch Slots` in gossip are updated every time a validator receives a
+   complete slot within the epoch. Completed slots are detected by blockstore
+   and sent over a channel to RepairService. It is important to note that we
+   know that by the time a slot `X` is complete, the epoch schedule must exist
+   for the epoch that contains slot `X` because WindowService will reject 
+   shreds for unconfirmed epochs.
+
+   Every `N/2` completed slots, the oldest `N/2` slots are moved from the 
+   `cache` into the `stash`. The base value `M` for the RLE should also
+   be updated.
+   
+## Repair Request Protocols
+
+The repair protocol makes best attempts to progress the forking structure of 
+Blockstore.
 
 The different protocol strategies to address the above challenges:
 
-1. Shred Repair \(Addresses Challenge \#1\): This is the most basic repair protocol, with the purpose of detecting and filling "holes" in the ledger. Blockstore tracks the latest root slot. RepairService will then periodically iterate every fork in blockstore starting from the root slot, sending repair requests to validators for any missing shreds. It will send at most some `N` repair reqeusts per iteration.
-
-   Note: Validators will only accept shreds within the current verifiable epoch \(epoch the validator has a leader schedule for\).
-
-2. Preemptive Slot Repair \(Addresses Challenge \#2\): The goal of this protocol is to discover the chaining relationship of "orphan" slots that do not currently chain to any known fork.
+1. Shred Repair \(Addresses Challenge \#1\): This is the most basic repair
+protocol, with the purpose of detecting and filling "holes" in the ledger.
+Blockstore tracks the latest root slot. RepairService will then periodically
+iterate every fork in blockstore starting from the root slot, sending repair
+requests to validators for any missing shreds. It will send at most some `N`
+repair reqeusts per iteration. Shred repair should prioritize repairing 
+forks with the most weight (ReplayStage communicates what the last heaviest
+fork was to RepairService). Validators should only send repair requests to 
+validators who have marked that slot as completed in their EpochSlots.
+
+   Note: Validators will only accept shreds within the current verifiable 
+   epoch \(epoch the validator has a leader schedule for\).
+
+2. Preemptive Slot Repair \(Addresses Challenge \#2\): The goal of this 
+protocol is to discover the chaining relationship of "orphan" slots that do not
```

Comment:
@carllin I don't think we need to track "orphans".  The protocol is much simpler.  Check epoch slots in gossip for anyone that has blocks that you don't have.  Ask them for packets.  We can prioritize orphans and anything near root and last working bank.

Reply:
The orphans protocol is already implemented, though there is no measure of prioritization right now. Maybe these should just be prioritized based on leader stake weight as well

Reply:
Let’s plan on sun-setting orphans once this is implemented.